mlir/tools/miopen-gen/miopen-gen.cpp

```diff
@@ -1195,248 +1204,136 @@ createVerifierFunc(ModuleOp &module, const KernelIF &kernel,
+  // obtain function name of the verifier wrapper
+  std::string verifyFuncName = "mcpuVerify5D";
+  if (valElemType.isF32())
+    verifyFuncName += "Float";
+  else
+    verifyFuncName += "Int32";
+
+  auto mr5DUnkType = MemRefType::get({-1, -1, -1, -1, -1}, valElemType);
+  bool isTestAndValSameType =
+      (testOutType.isInteger(32) || testOutType.isF32());
+
+  mlir::Value testResult, valResult; // Values passed to the verify function
+  mlir::Value testResultNew, valResultTmp; // Values used for type conversion
+  if (!isTestAndValSameType) {
+    // When gpu kernel output data type = f16 | bf16, type conversions
+    // are required before calling the verify function
+
+    // step 1: cast test result to the same type as valid result
+
+    // clang-format off
+    // %0 = memref.alloc() : memref<1x64x3x7x7xf32>
+    // %1 = memref.collapse_shape %arg0 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf16> into memref<9408xf16>
+    // %2 = memref.cast %1 : memref<9408xf16> to memref<?xf16>
+    // %3 = memref.collapse_shape %0 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf32> into memref<9408xf32>
+    // %4 = memref.cast %3 : memref<9408xf32> to memref<?xf32>
+    // %c9408 = arith.constant 9408 : index
+    // call @_memcpy_f16_f32(%2, %4, %c9408) : (memref<?xf16>, memref<?xf32>, index) -> ()
+    // clang-format on
+
+    testResultNew = b.create<memref::AllocOp>(loc, valType);
+    emitMemcpy(b, arg0, testResultNew);
+
+    // step 2: cast valid result down to the same type as test result and cast
+    // back
+    //   For f16 and bf16 datatypes, gpu hardware outputs f32 results, which are
+    //   truncated to f16/bf16 before returning from the gpu kernel
+    //   To make the comparison fair, the truncation step is added manually to
+    //   the validation results.
+
+    // clang-format off
+    // %5 = memref.alloc() : memref<1x64x3x7x7xf16>
+    // %6 = memref.collapse_shape %arg1 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf32> into memref<9408xf32>
+    // %7 = memref.cast %6 : memref<9408xf32> to memref<?xf32>
+    // %8 = memref.collapse_shape %5 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf16> into memref<9408xf16>
+    // %9 = memref.cast %8 : memref<9408xf16> to memref<?xf16>
+    // %c9408_2 = arith.constant 9408 : index
+    // call @_memcpy_f32_f16(%7, %9, %c9408_2) : (memref<?xf32>, memref<?xf16>, index) -> ()
+    // %10 = memref.collapse_shape %5 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf16> into memref<9408xf16>
+    // %11 = memref.cast %10 : memref<9408xf16> to memref<?xf16>
+    // %12 = memref.collapse_shape %arg1 [[0, 1, 2, 3, 4]] : memref<1x64x3x7x7xf32> into memref<9408xf32>
+    // %13 = memref.cast %12 : memref<9408xf32> to memref<?xf32>
+    // %c9408_3 = arith.constant 9408 : index
+    // call @_memcpy_f16_f32(%11, %13, %c9408_3) : (memref<?xf16>, memref<?xf32>, index) -> ()
+    // clang-format on
+
+    valResultTmp = b.create<memref::AllocOp>(loc, testType);
```

Comment:
Ok, here, I don't think we need to allocate. Why not instead loop through the test result, cast each element to the narrower type, then cast it back. `extf` and `truncf` should do it. (along with `createAffineLoopNest`)

Reply:
How about reusing `arg0` to hold the narrower type of the validation result?
```cpp
emitMemcpy(b, arg1, arg0);
emitMemcpy(b, arg0, arg1)
```
Since `arg0`'s values are copied (and extended) to `testResultNew`, it should be safe to overwrite it.

Reply:
Yeah, it should be safe, but that sort of aliasing feels like a trap that'll come back to bite us when we forgot we did it. It feels like it'd be simpler to write an in-place elementwise truncate + extend. Heck, you could probably use a `linalg.generic` for it.

(and part of the reason I was wanting to cut down on extra buffers was that these buffers are already on the large side)